### PR TITLE
[FEATURE] Improve test coverage for MLHttpClientFactory.java

### LINKED ISSUE #1379: [FEATURE] Improve test coverage for MLHttpClientFactory.java
We need to improve test coverage in our repo.

This issue will target only to improve test coverage for [MLHttpClientFactory.java class](https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactory.java)

From this url we can see where test coverage is missing for this class: https://app.codecov.io/gh/opensearch-project/ml-commons/blob/2.x/ml-algorithms%2Fsrc%2Fmain%2Fjava%2Forg%2Fopensearch%2Fml%2Fengine%2Fhttpclient%2FMLHttpClientFactory.java

### PR DESCRIPTION
**Pull request by akolarkunnu**
### Description
Improved the test coverage by adding edge cases.
The if check (bytes.length != 4) is an unreachable code, Inet4Address always will have array of 4 bytes address.

### Related Issues
Resolves #1379

### Check List
- [x] New functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by akolarkunnu** [comment]
Code Coverage report after fix:

**Comment by akolarkunnu** [comment]
linux (21) - Known flaky org.opensearch.ml.rest.RestMLRemoteInferenceIT.testPredictWithAutoDeployAndTTL_RemoteModel #3544

Windows (23) - This is a test only change, failures are not related to the code changes.

**Comment by akolarkunnu** [comment]
Hi Maintainers, Any comments, please.

**Comment by dhrubo-os** [comment]
```
Tests with failures:
 - org.opensearch.ml.rest.RestMLRemoteInferenceIT.testPredictWithAutoDeployAndTTL_RemoteModel
 ```
 
 Can you take updates from main?

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java [hunk 1]
```diff
@@ -29,84 +35,115 @@ public void test_getSdkAsyncHttpClient_success() {
...
-        MLHttpClientFactory.validate("http", "0:0:0:0:0:ffff:127.0.0.1", 80, privateIpEnabled);
-        MLHttpClientFactory.validate("http", "153.24.76.232", 80, privateIpEnabled);
+        MLHttpClientFactory.validate(HTTP, "0254.020.00.01", 80, PRIVATE_IP_ENABLED);
+        MLHttpClientFactory.validate(HTTP, "172.1048577", 80, PRIVATE_IP_ENABLED);
+        MLHttpClientFactory.validate(HTTP, "2886729729", 80, PRIVATE_IP_ENABLED);
+        MLHttpClientFactory.validate(HTTP, "192.11010049", 80, PRIVATE_IP_ENABLED);
+        MLHttpClientFactory.validate(HTTP, "3232300545", 80, PRIVATE_IP_ENABLED);
+        MLHttpClientFactory.validate(HTTP, "0:0:0:0:0:ffff:127.0.0.1", 80, PRIVATE_IP_ENABLED);
+        MLHttpClientFactory.validate(HTTP, "153.24.76.232", 80, PRIVATE_IP_ENABLED);
+        MLHttpClientFactory.validate(HTTP, "10.24.76.186", 80, PRIVATE_IP_ENABLED);
```
Comment (dhrubo-os) [inline ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java:112]: should we add test for https too here?

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java [hunk 2]
```diff
@@ -29,84 +35,115 @@ public void test_getSdkAsyncHttpClient_success() {
...
         try {
-            MLHttpClientFactory.validate("http", "3232300545", 80, privateIpEnabled);
+            MLHttpClientFactory.validate(HTTP, "153.24.76.232", 80, PRIVATE_IP_DISABLED);
         } catch (IllegalArgumentException e) {
             assertNotNull(e);
         }
 
         try {
-            MLHttpClientFactory.validate("http", "0:0:0:0:0:ffff:127.0.0.1", 80, privateIpEnabled);
+            MLHttpClientFactory.validate(HTTP, "177.0.0.1", 80, PRIVATE_IP_DISABLED);
```
Comment (dhrubo-os) [inline ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java:91]: why are we removing these lines?

**Comment by akolarkunnu** [comment]
> ```
> Tests with failures:
>  - org.opensearch.ml.rest.RestMLRemoteInferenceIT.testPredictWithAutoDeployAndTTL_RemoteModel
> ```
> 
> Can you take updates from main?

Done.

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java [hunk 3]
```diff
@@ -36,7 +36,9 @@ public void test_getSdkAsyncHttpClient_success() {
     @Test
     public void test_validateIp_validIp_noException() throws Exception {
         MLHttpClientFactory.validate(HTTP, TEST_HOST, 80, PRIVATE_IP_DISABLED);
+        MLHttpClientFactory.validate(HTTPS, TEST_HOST, 80, PRIVATE_IP_DISABLED);
```
Comment (dhrubo-os) [inline ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java:39]: HTTPS should be 443, why it's allowing 80 here?

Comment (akolarkunnu) [reply ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java:39]: If we pass port as -1, then it will reset to 443, otherwise it accepts all the port values between 0-65536 (according to validation logic in the code).

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java (hunk 2 above)
Comment (akolarkunnu) [reply ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java:93]: If you look at closely, we are not removing any lines here. git diff shows a little weird here.

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java (hunk 1 above)
Comment (akolarkunnu) [reply ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java:116]: Only different logic for https is the resetting port number to 443 if user pass it as -1. Already there is a specific test case to handle this scenario.
Still I added 4 new test cases to cover valid and rare ip addresses of https.

**Comment by dhrubo-os** [comment]
```
* What went wrong:
Execution failed for task ':opensearch-ml-plugin:compileTestJava'.
> Compilation failed; see the compiler output below.
  Note: Some input files use unchecked or unsafe operations.
  Note: Recompile with -Xlint:unchecked for details.
  Note: Recompile with -Xlint:deprecation for details.
  Note: Some input files use or override a deprecated API.
  /__w/ml-commons/ml-commons/plugin/src/test/java/org/opensearch/ml/tools/PromptHandler.java:18: error: cannot find symbol
          return prompt.contains(llmThought().getQuestion());
                                             ^
    symbol:   method getQuestion()
    location: class LLMThought
  /__w/ml-commons/ml-commons/plugin/src/test/java/org/opensearch/ml/tools/PromptHandler.java:35: error: cannot find symbol
                  + this.llmThought().getAction()
                                     ^
    symbol:   method getAction()
    location: class LLMThought
  /__w/ml-commons/ml-commons/plugin/src/test/java/org/opensearch/ml/tools/PromptHandler.java:38: error: cannot find symbol
                  + this.llmThought().getActionInput()
                                     ^
    symbol:   method getActionInput()
    location: class LLMThought
  /__w/ml-commons/ml-commons/plugin/src/test/java/org/opensearch/ml/tools/MockLLM.java:40: error: cannot find symbol
                      llmResponse.setCompletion(promptHandler.response(prompt));
                                 ^
    symbol:   method setCompletion(String)
    location: variable llmResponse of type LLMResponse
  /__w/ml-commons/ml-commons/plugin/src/test/java/org/opensearch/ml/tools/VisualizationsToolIT.java:36: error: cannot find symbol
                      .builder()
                      ^
    symbol:   method builder()
    location: class LLMThought
  /__w/ml-commons/ml-commons/plugin/src/test/java/org/opensearch/ml/tools/VisualizationsToolIT.java:46: error: cannot find symbol
                      .builder()
                      ^
    symbol:   method builder()
    location: class LLMThought
  /__w/ml-commons/ml-commons/plugin/src/test/java/org/opensearch/ml/utils/TestHelper.java:15: error: cannot find symbol
  import static org.opensearch.cluster.node.DiscoveryNodeRole.SEARCH_ROLE;
  ^
    symbol:   static SEARCH_ROLE
    location: class DiscoveryNodeRole

```

This [PR](https://github.com/opensearch-project/ml-commons/pull/3667) is merged. Please rebase again.

**Comment by akolarkunnu** [comment]
> This [PR](https://github.com/opensearch-project/ml-commons/pull/3667) is merged. Please rebase again.

Ok, done.

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java (hunk 3 above)
Comment (akolarkunnu) [reply ml-algorithms/src/test/java/org/opensearch/ml/engine/httpclient/MLHttpClientFactoryTests.java]: Also changed all HTTPS ports to 443 now

**Review by mingshl — APPROVED** [review]

**Comment by dhrubo-os** [comment]
Thanks @akolarkunnu for working on this. Merging the PR.
